### DISCUSSION TITLE
How to properly specify multiple `belongs_to()` rules pointing to same table?

### DISCUSSION THREAD

**Discussion by kamakazikamikaze** [emmett-framework/emmett#500] [Q&A] [answered]
I've hit an issue defining a table where a row contains two foreign keys from the same target table.

For example:

```python

class User(Model):
    primary_keys = ["user_id"]
    user_id = Field.int()
    name = Field().string()
    birthday = Field.datetime()
    has_many({"sent": "Whisper.sender_user_id"}, {"received": "Whisper.receiver_user_id"})

class Whisper(Model):
    belongs_to({"sender": "User"}, {"receiver": "User"})
    message = Field.text()
    timestamp = Field.datetime()
```

Running this will raise an error of `psycopg2.error.InvalidForeignKey: foreign key referenced-columns list must not contain duplicates`

It seems to stem from something similar to this [StackOverflow post](https://stackoverflow.com/questions/62814926/postgresql-composite-foreign-key-columns-list-must-not-contain-duplicates). Apparently the foreign keys need to be announced separately instead of all at the same time? Is that something I can do?

**Comment by gi0baro** [comment]
@kamakazikamikaze the `has_many` definition seems wrong, should look like this:

```python
has_many({"sent": "Whisper.sender"}, {"received": "Whisper.receiver"})
```

I also don't really get why you would define a `User` table with an integer primary key `user_id` instead of using the default id implementation..

**Reply by kamakazikamikaze** [reply]
The actual database schema I'm using is different but my workplace requires us to change any code we share so there's no issue with leaking intellectual property. I wrote this real quick. I think the schema would be closer to the following:

```python

class User(Model):
    primary_keys = ["last_name", "first_name"]
    first_name = Field().string(length=32)
    last_name = Field().string(length=32)
    birthday = Field.datetime()
    has_many({"sent": "Whisper.sender"}, {"received": "Whisper.receiver"})

class Whisper(Model):
    belongs_to({"sender": "User"}, {"receiver": "User"})
    message = Field.text()
    timestamp = Field.datetime()
```

Uniqueness is established on two string fields for `User`. Both `Whisper.sender` and `Whisper.receiver` must be foreign keys from `User`.

The migration is having an issue defining the foreign key. The auto-generated code looks something like this:

`self.create_foreign_key('whispers_ecnt__fk_...', 'whispers', 'users', ['sender_first_name', 'sender_last_name', 'receiver_first_name', 'receiver_last_name'], ['first_name', 'last_name', 'first_name', 'last_name'], on_delete='CASCADE')`

**Reply by gi0baro** [reply]
Now I get it. So it seems ORM right now doesn't support this.
I might be able to check the migrations code in the following days; it might be just a generation issue.

If you split the generated condition into two different commands, one per `belongs_to` item, does it work? The next migration generation will probably revert the thing, but at least we can confirm this is just a migration issue.

**Reply by kamakazikamikaze** [reply]
Tried it this morning. I'm hitting the same error despite splitting the `belongs_to` commands.

Right now this is just a proof-of-concept project intended to match the existing design of something already deployed using a different framework. No prod data is being used here and the database is being rebuilt via Docker each run, so I can perform as many tests as you'd like to help narrow things down.

**Reply by gi0baro** [reply]
@kamakazikamikaze can you please check if https://github.com/emmett-framework/emmett/pull/501 fixes your issue?

**Reply by kamakazikamikaze** [reply]
I'll give it a shot. Stay tuned.

**Reply by kamakazikamikaze** [reply]
It works! 🎉

**Reply by gi0baro** [reply, marked as answer]
👍 I'll release a patch version the next week
